pkg/agent/stats/collector.go

```diff
@@ -99,16 +102,12 @@ func (m *Collector) Run(stopCh <-chan struct{}) {
 // collect collects the stats of Openflow rules, maps them to the stats of NetworkPolicies.
 // It returns a map from NetworkPolicyReferences to their stats.
 func (m *Collector) collect() *statsCollection {
-	// TODO: The following process is not atomic, there's a chance that the ofID is released and reused by another
-	//  NetworkPolicy rule in-between, leading to incorrect metrics. We should return relevant NetworkPolicy references
-	//  along with metrics to avoid it.
 	ruleStatsMap := m.ofClient.NetworkPolicyMetrics()
 	npStatsMap := map[types.UID]*statsv1alpha1.TrafficStats{}
 	acnpStatsMap := map[types.UID]*statsv1alpha1.TrafficStats{}
 	anpStatsMap := map[types.UID]*statsv1alpha1.TrafficStats{}
 	for ofID, ruleStats := range ruleStatsMap {
-		policyRef := m.ofClient.GetPolicyFromConjunction(ofID)
-		// Same as above, this may be because the NetworkPolicy is removed right after the metrics are fetched.
+		policyRef := m.networkPolicyQuerier.GetNetworkPolicyByRuleFlowID(ofID)
 		if policyRef == nil {
 			klog.Infof("Cannot find NetworkPolicy that has ofID %v", ofID)
```

Comment:
maybe this should become a warning now?

also I would recommend adding a comment to explain that this should not happen because the id to rule mapping is preserved for 5 seconds after an id is released

Reply:
Agreed. Added the comment.